File: packages/api-types/src/messages.ts
import { z } from "zod";

/**
 * `apps/server/src/cocktail_server/schemas/messages.py` と同じ判別 union を
 * クライアント側 Zod で再現する。サーバとキー名・制約をそろえる責務があるので
 * 変更するときは両側をセットで更新すること。
 */

const UUID_PATTERN = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/;

export const RoleSchema = z.enum(["user", "assistant", "tool", "system"]);
export type Role = z.infer<typeof RoleSchema>;

export const ToolCallStatusSchema = z.enum(["pending", "running", "done", "error"]);
export type ToolCallStatus = z.infer<typeof ToolCallStatusSchema>;

export const TextPartSchema = z
  .object({
    type: z.literal("text"),
    text: z.string().min(1).max(10_000),
  })
  .strict();
export type TextPart = z.infer<typeof TextPartSchema>;

export const ImagePartSchema = z
  .object({
    type: z.literal("image"),
    image_id: z.string().regex(UUID_PATTERN),
    mime: z.string().min(1).max(64),
    width: z.number().int().min(1).max(8192).nullable().optional(),
    height: z.number().int().min(1).max(8192).nullable().optional(),
  })
  .strict();
export type ImagePart = z.infer<typeof ImagePartSchema>;

export const ToolCallPartSchema = z
  .object({
    type: z.literal("tool_call"),
    id: z.string().min(1).max(128),
    name: z.string().min(1).max(64),
    args: z.record(z.unknown()),
    status: ToolCallStatusSchema,
  })
  .strict();
export type ToolCallPart = z.infer<typeof ToolCallPartSchema>;

export const ToolResultPartSchema = z
  .object({
    type: z.literal("tool_result"),
    call_id: z.string().min(1).max(128),
    summary: z.string().max(1000),
    data: z.record(z.unknown()),
  })
  .strict();
export type ToolResultPart = z.infer<typeof ToolResultPartSchema>;

export const ContentPartSchema = z.discriminatedUnion("type", [
  TextPartSchema,
  ImagePartSchema,
  ToolCallPartSchema,
  ToolResultPartSchema,
]);
export type ContentPart = z.infer<typeof ContentPartSchema>;

export const UserContentPartSchema = z.discriminatedUnion("type", [
  TextPartSchema,
  ImagePartSchema,
]);
export type UserContentPart = z.infer<typeof UserContentPartSchema>;

export const MessageSchema = z
  .object({
    id: z.string().min(1).max(128),
    conversation_id: z.string().min(1).max(128),
    role: RoleSchema,
    parts: z.array(ContentPartSchema).min(1).max(32),
    created_at: z.string(),
    parent_id: z.string().max(128).nullable().optional(),
  })
  .strict();
export type Message = z.infer<typeof MessageSchema>;

File: apps/server/src/cocktail_server/schemas/messages.py
from __future__ import annotations

from datetime import datetime
from typing import Annotated, Any, Literal

from pydantic import BaseModel, ConfigDict, Field

_UUID_PATTERN = r"^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$"

Role = Literal["user", "assistant", "tool", "system"]
ToolCallStatus = Literal["pending", "running", "done", "error"]


class TextPart(BaseModel):
    """テキスト。ユーザ入力でも、アシスタント出力の自然言語でも使う。"""

    model_config = ConfigDict(strict=True, extra="forbid")

    type: Literal["text"] = "text"
    text: str = Field(min_length=1, max_length=10_000)


class ImagePart(BaseModel):
    """画像参照。`image_id` は `POST /images` か `generate_image` ツールが発行した UUID。"""

    model_config = ConfigDict(strict=True, extra="forbid")

    type: Literal["image"] = "image"
    image_id: str = Field(pattern=_UUID_PATTERN)
    mime: str = Field(min_length=1, max_length=64)
    width: int | None = Field(default=None, ge=1, le=8192)
    height: int | None = Field(default=None, ge=1, le=8192)


class ToolCallPart(BaseModel):
    """アシスタントがツールを呼び出した記録。`args` は OpenAPI の付随ペイロード。"""

    model_config = ConfigDict(strict=True, extra="forbid")

    type: Literal["tool_call"] = "tool_call"
    id: str = Field(min_length=1, max_length=128)
    name: str = Field(min_length=1, max_length=64)
    args: dict[str, Any]
    status: ToolCallStatus


class ToolResultPart(BaseModel):
    """ツールの結果。画像などの成果物は別途 `ImagePart` として同メッセージに同居させる。"""

    model_config = ConfigDict(strict=True, extra="forbid")

    type: Literal["tool_result"] = "tool_result"
    call_id: str = Field(min_length=1, max_length=128)
    summary: str = Field(max_length=1000)
    data: dict[str, Any]


ContentPart = Annotated[
    TextPart | ImagePart | ToolCallPart | ToolResultPart,
    Field(discriminator="type"),
]
"""アシスタント/ツールメッセージで許容される全パート。"""

UserContentPart = Annotated[
    TextPart | ImagePart,
    Field(discriminator="type"),
]
"""ユーザ入力で許容されるパート（tool_call/result はサーバ生成のみ）。"""


class Message(BaseModel):
    """1 ターン 1 メッセージ。`parts` の並びが UI の表示順になる。"""

    model_config = ConfigDict(strict=True, extra="forbid")

    id: str = Field(min_length=1, max_length=128)
    conversation_id: str = Field(min_length=1, max_length=128)
    role: Role
    parts: list[ContentPart] = Field(min_length=1, max_length=32)
    created_at: datetime
    parent_id: str | None = Field(default=None, max_length=128)
